### PR TITLE
[5.8] Job-based retry delay

### PR DESCRIPTION
**Pull request by bahdcoder**
#### Related links
This pull request implements the feature requested  [here](https://github.com/laravel/ideas/issues/537)

#### What does this feature do?
If a `$retriesIn` property is defined on a job, then if that job fails the retry will be delayed for `$retriesIn` amount of time.

#### How will this feature be useful to users?
Let's take an example. An application that  creates `Digital Ocean` droplets using the API. 

The application then tries to `SSH` into the droplet using a job `SshIntoDroplet` once it's active. Even though the droplet is marked as `active` sometimes `SSH` is just not ready, and the job fails. 

With the current functionality, this job is retried immediately. This probably isn't the best way to handle this specific job. 

To fix this, this pull request adds a `$retriesIn` property on a job, and the job is only re-attempted after the defined amount of time.

#### How to use?

Here's an example job that implements this feature:

```php
<?php

namespace App\Jobs;

use App\Droplet;
use Illuminate\Bus\Queueable;
use Illuminate\Queue\SerializesModels;
use Illuminate\Queue\InteractsWithQueue;
use Illuminate\Contracts\Queue\ShouldQueue;
use Illuminate\Foundation\Bus\Dispatchable;

class SshIntoDroplet implements ShouldQueue
{
    use Dispatchable, InteractsWithQueue, Queueable, SerializesModels;

    /**
     * The droplet to ssh into.
     *
     * @var Droplet
     */
    public $droplet;

    /**
     * The delay before attempting job again
     * 
     * @var integer
     */
    public $retriesIn = 5;

    /**
     * Create a new job instance.
     *
     * @return void
     */
    public function __construct(Droplet $droplet)
    {
        $this->droplet = $droplet;
    }

    /**
     * Execute the job.
     *
     * @return void
     */
    public function handle()
    {
        // job fails when droplet is active but still booting.
        // If it fails, the Queue worker waits for 5 seconds before retrying.
        $this->droplet->ssh();
    }
}

```

### CODE REVIEW

File: src/Illuminate/Queue/Worker.php [hunk 1]
```diff
@@ -321,6 +321,11 @@ public function process($connectionName, $job, WorkerOptions $options)
                 $connectionName, $job, (int) $options->maxTries
             );
 
+            // Here we will check if there's a delay before retrying a job.
+            $this->delayJobAttemptExecution(
+              $job
+            );
```
Comment (driesvints) [inline src/Illuminate/Queue/Worker.php]: Please put this on one line.

File: src/Illuminate/Queue/Worker.php [hunk 2]
```diff
@@ -482,6 +487,25 @@ protected function raiseExceptionOccurredJobEvent($connectionName, $job, $e)
         ));
     }
 
+    /**
+     * Delay a job retry if defined.
+     *
+     * @param  \Illuminate\Contracts\Queue\Job  $job
+     * @return  void
```
Comment (driesvints) [inline src/Illuminate/Queue/Worker.php]:
```suggestion
     * @return void
```

**Review by driesvints — CHANGES_REQUESTED** [review]
Please fix the tests as well.

**Comment by bahdcoder** [comment]
Hello @driesvints . Could you have a look at the failing tests? They seem to fail on the `5.8` branch even without my implementation.

**Comment by bahdcoder** [comment]
Hello @GrahamCampbell for some strange reason, running the tests on a freshly new cloned copy of the 5.8 branch locally fails. Same thing I'm experiencing on this branch, and I have no idea why. Could you help out? Thanks.

**Comment by benignant** [comment]
Quite a few people are interested in this. What happened?

**Comment by rokde** [comment]
This would be very helpful.

**Comment by taylorotwell** [comment]
I'm reviving this. Should be done today.

**Comment by bahdcoder** [comment]
Hey @taylorotwell I initially opened this pull request. Should I reopen? Do you have any improvements you would love me to implement, or are you working on it?

**Comment by devcircus** [comment]
Another pr has  already been merged.

**Comment by devcircus** [comment]
#28265

**Pull request closed without merging**